javalib/src/main/scala/java/util/HashMap.scala

```diff
@@ -12,153 +12,502 @@
 
 package java.util
 
-import scala.collection.mutable
+import java.{util => ju}
 
-class HashMap[K, V] protected (inner: mutable.Map[Box[K], V])
+import ScalaOps._
+
+class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
     extends AbstractMap[K, V] with Serializable with Cloneable {
   self =>
 
-  def this() =
-    this(mutable.HashMap.empty[Box[K], V])
+  import HashMap._
 
-  def this(initialCapacity: Int, loadFactor: Float) = {
-    this()
-    if (initialCapacity < 0)
-      throw new IllegalArgumentException("initialCapacity < 0")
-    else if (loadFactor < 0.0)
-      throw new IllegalArgumentException("loadFactor <= 0.0")
-  }
+  if (initialCapacity < 0)
+    throw new IllegalArgumentException("initialCapacity < 0")
+  if (loadFactor <= 0.0)
+    throw new IllegalArgumentException("loadFactor <= 0.0")
+
+  def this() =
+    this(HashMap.DEFAULT_INITIAL_CAPACITY, HashMap.DEFAULT_LOAD_FACTOR)
 
   def this(initialCapacity: Int) =
     this(initialCapacity, HashMap.DEFAULT_LOAD_FACTOR)
 
   def this(m: Map[_ <: K, _ <: V]) = {
-    this()
+    this(m.size())
     putAll(m)
   }
 
-  override def clear(): Unit =
-    inner.clear()
+  /** The actual hash table.
+   *
+   *  In each bucket, nodes are sorted by increasing value of `hash`.
+   */
+  private[this] var table = new Array[Node[K, V]](tableSizeFor(initialCapacity))
```

Comment:
It turns out that it is *essential* to adjust `initialCapacity` with `tableSizeFor` in all cases, because the length of the table must always be a power of 2. The algorithms rely on this, notably the formula in `def index`. So I have reverted to the previous state.

Reply:
OK. Fair enough. That's what I thought. I agree that keeping the growing simple is more important than following the spec to the word. However, could we add a comment / document this discrepancy and why?

Reply:
Yes, good point. I'll add a comment.

Reply:
I can't find the comment :S

Reply:
Oops, indeed :s Now it's there.